### PR TITLE
feat(query): Procedure Part2 support arguments

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Parat II of https://github.com/datafuselabs/databend/pull/16348

this pr support call procedure(<args>);

After this pr, we need to add a new issue to enhanced procedure checks to make it more of a product than a prototype

fixes: Part of https://github.com/datafuselabs/databend/issues/14904

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16453)

### CODE REVIEW

**Review by drmingdrmer — COMMENTED** [review]
> Reviewed 12 of 24 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16453)* status: 12 of 24 files reviewed, 2 unresolved discussions (waiting on @TCeason)

___
*[`src/meta/app/src/principal/procedure.rs` line 76 at r1](https://reviewable.io/reviews/datafuselabs/databend/16453#-O6iaUKk3t9GXqm1XOnQ:-O6iaUKk3t9GXqm1XOnR:b-e20pjw) ([raw file](https://github.com/datafuselabs/databend/blob/3612083dc43cf098726251bed16aefbda7ffd381/src/meta/app/src/principal/procedure.rs#L76)):*
> ```XML
>             "Lanuage: {:?}, args {:?} return_type: {:?}, CreatedOn: {:?}, Script: {:?}, Comment: {:?}",
>             self.procedure_language,
>             self.arg_names,
> ```

use `DisplaySliceExt::display_n::<1000>()` to build better for human output.

Other fields can also be benefitted from this method.



_Suggestion:_
```XML
            "Lanuage: {:?}, args {} return_type: {:?}, CreatedOn: {:?}, Script: {:?}, Comment: {:?}",
            self.procedure_language,
            self.arg_names.display_n::<1000>(),
```
___
*[`src/meta/app/src/principal/procedure_identity.rs` line 36 at r1](https://reviewable.io/reviews/datafuselabs/databend/16453#-O6ib08C6ONb8py6Bz75:-O6ib08C6ONb8py6Bz76:b-7pyttc) ([raw file](https://github.com/datafuselabs/databend/blob/3612083dc43cf098726251bed16aefbda7ffd381/src/meta/app/src/principal/procedure_identity.rs#L36)):*
> ```XML
>         Self {
>             name: name.to_string(),
>             args: args.to_string().to_lowercase(),
> ```

Why is `args` special and is converted to lower cases?

**Comment by TCeason** [comment]
Why is args special and is converted to lower cases?

No need to modify it. Already convert to lower in parse create/drop procedure.

e.g.

create procedure p1(x UInt8) -> should can be drop procedure p1(uint8)

**Comment by TCeason** [comment]
warit https://github.com/datafuselabs/databend/pull/16455 merged.

File: src/query/service/src/interpreters/util.rs [hunk 1]
```diff
@@ -61,3 +84,300 @@ pub fn generate_desc_schema(
...
+            0 => DataBlock::empty_with_schema(schema.clone()),
+            1 => blocks[0].clone(),
+            _ => DataBlock::concat(&blocks)?,
+        };
+
+        Ok(QueryResult { schema, block })
+    }
+
+    fn var_to_ast(&self, scalar: &Self::Var) -> databend_common_exception::Result<Expr> {
+        let ast = match scalar {
```
Comment (sundy-li) [inline src/query/service/src/interpreters/util.rs:126]: we can use `scalar.sql_display` and parse it back

File: tests/sqllogictests/suites/base/15_procedure/15_0002_procedure.test [hunk 2]
```diff
@@ -49,17 +44,27 @@ call procedure p1();
...
+call procedure p1(1,10);
+----
+20
+
 query T
 select name, arguments from system.procedures where name = 'p1';
 ----
 p1 p1() RETURN (Int32)
-p1 p1(Int32,Int32) RETURN (Int32)
+p1 p1(uint8,uint8) RETURN (Int32)
```
Comment (sundy-li) [inline tests/sqllogictests/suites/base/15_procedure/15_0002_procedure.test:61]: lower case ?

Comment (TCeason) [reply tests/sqllogictests/suites/base/15_procedure/15_0002_procedure.test:61]: Yes. In this pr I make the type name as lower case when parse.

https://github.com/datafuselabs/databend/pull/16453/files#diff-391b369d101e47b8108e9a90993e8a08f926edd73ddc5a2d0c2bed1a4bf90703R2079-R2082
https://github.com/datafuselabs/databend/pull/16453/files#diff-391b369d101e47b8108e9a90993e8a08f926edd73ddc5a2d0c2bed1a4bf90703R2149-R2150

e.g.

create procedure p1(x UInt8) -> should can be drop procedure p1(uint8)

File: src/query/service/src/interpreters/util.rs (hunk 1 above)
Comment (TCeason) [reply src/query/service/src/interpreters/util.rs:126]: It's refactored from src/query/service/src/interpreters/interpreter_execute_immediate.rs. cc @andylokandy Cloud we replace it as sql_display?

File: tests/sqllogictests/suites/base/15_procedure/15_0002_procedure.test (hunk 2 above)
Comment (TCeason) [reply tests/sqllogictests/suites/base/15_procedure/15_0002_procedure.test]: Done

File: src/query/service/src/interpreters/util.rs (hunk 1 above)
Comment (andylokandy) [reply src/query/service/src/interpreters/util.rs:126]: May use `ScalarRef::to_string` and then `parse_expr` it back. But, currently, it's not guaranteed that it must success.

File: src/query/service/src/interpreters/util.rs [hunk 3]
```diff
@@ -61,3 +84,300 @@ pub fn generate_desc_schema(
...
+                    row,
+                    set.block.num_rows(),
+                ))
+            })?
+            .to_owned();
+
+        Ok(cell)
+    }
+
+    fn set_len(&self, set: &Self::Set) -> usize {
```
Comment (andylokandy) [inline src/query/service/src/interpreters/util.rs:371]: maybe better rename to `num_rows`. bad decision before XD

File: src/query/service/src/interpreters/interpreter_procedure_call.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            let tokens = tokenize_sql(&self.plan.script)?;
+            let mut ast = run_parser(
+                &tokens,
+                sql_dialect,
+                ParseMode::Template,
+                false,
+                script_block,
+            )?;
+
+            for declare in ast.declares {
```
Comment (andylokandy) [inline src/query/service/src/interpreters/interpreter_procedure_call.rs:88]: May add tests for ScripBlock.declares

Comment (TCeason) [reply src/query/service/src/interpreters/interpreter_procedure_call.rs:88]: You mean add test like this:
```sql
CREATE PROCEDURE p1(x UInt8, sum UInt8) RETURNS int not null LANGUAGE SQL COMMENT='test' AS $$
declare x := 20;
BEGIN
    FOR x IN x TO x + 3 DO
        sum := sum + x;
    END FOR;
    RETURN sum;
END;
$$;
```

I think this is a check and we need to do other check like this in next pr to enhance procedure.

File: src/query/service/src/interpreters/util.rs (hunk 3 above)
Comment (TCeason) [reply src/query/service/src/interpreters/util.rs]: Done

**Review by drmingdrmer — APPROVED** [review]
> Reviewed 8 of 24 files at r1, 1 of 3 files at r2, 3 of 4 files at r3, 7 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16453)* status: 24 of 28 files reviewed, 4 unresolved discussions (waiting on @andylokandy and @sundy-li)

File: src/query/service/src/interpreters/util.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/service/src/interpreters/util.rs:126]: `sql_display`  handles inf/nan better.

Comment (sundy-li) [reply src/query/service/src/interpreters/util.rs:126]: see `parse_computed_expr_to_string`

Comment (TCeason) [reply src/query/service/src/interpreters/util.rs]: Done

**Review by sundy-li — APPROVED** [review]

**Comment by sundy-li** [comment]
docs  cc@soyeric128 with @TCeason

**Pull request merged**